### ISSUE TITLE
ZefyrField can not remove the last character on dev-1.0

### ISSUE THREAD

**Issue by khanhnguyen-96** [memspace/zefyr#480]
## Steps to Reproduce

NOTE: This issue is found on dev-1.0.

1.  Load an empty Notus document (I can reproduce this in the original Zefyr example code, as long as the loaded document is empty which means it  only contains "\n")
2. Try to type in a few words like "abc abc babc abc", but not too long
3. Try to remove all the content by pressing/holding backspace
4. Pressing backspace can not delete the last character

Demo:
![bug](https://user-images.githubusercontent.com/27147322/108301062-d32e5800-71d3-11eb-92da-5a2dfbc9ba12.gif)

## Logs

```
No specific log
```

```
[✓] Flutter (Channel stable, 1.22.6, on Mac OS X 10.15.6 19G73 darwin-x64, locale en)
    • Flutter version 1.22.6 at /Users/teq-ios/development/flutter
    • Framework revision 9b2d32b605 (2 weeks ago), 2021-01-22 14:36:39 -0800
    • Engine revision 2f0af37152
    • Dart version 2.10.5

 
[!] Android toolchain - develop for Android devices (Android SDK version 30.0.2)
    • Android SDK at /Users/teq-ios/Library/Android/sdk
    • Platform android-30, build-tools 30.0.2
    • ANDROID_HOME = /Users/teq-ios/Library/Android/sdk
    • Java binary at: /Applications/Android Studio.app/Contents/jre/jdk/Contents/Home/bin/java
    • Java version OpenJDK Runtime Environment (build 1.8.0_242-release-1644-b3-6915495)
    ! Some Android licenses not accepted.  To resolve this, run: flutter doctor --android-licenses

[✓] Xcode - develop for iOS and macOS (Xcode 12.4)
    • Xcode at /Applications/Xcode.app/Contents/Developer
    • Xcode 12.4, Build version 12D4e
    • CocoaPods version 1.10.0

[!] Android Studio (version 4.1)
    • Android Studio at /Applications/Android Studio.app/Contents
    ✗ Flutter plugin not installed; this adds Flutter specific functionality.
    ✗ Dart plugin not installed; this adds Dart specific functionality.
    • Java version OpenJDK Runtime Environment (build 1.8.0_242-release-1644-b3-6915495)

[✓] Connected device (1 available)
    • iPhone X (mobile) • 0E7083D8-7982-4E96-9DB7-67D5085D2094 • ios • com.apple.CoreSimulator.SimRuntime.iOS-12-4 (simulator)

! Doctor found issues in 2 categories.

```

**Comment by namtranx07** [comment]
I had the same problem. Please fix that

**Comment by singerdmx** [comment]
Is it mobile or web

**Comment by tantai240495** [comment]
The same issue goes here, mine is on Android. 
Device: Redmi K20 Pro (Android 10)
Any suggestion for this?

**Comment by singerdmx** [comment]
Is it similar to https://github.com/memspace/zefyr/issues/478

**Comment by nnkti18** [comment]
> Is it mobile or web

Mobile (both Android and iOS), I haven't tested this on web because I have no intention of developing my app on web.

> Is it similar to #478
No, I don't think it is

**Comment by singerdmx** [comment]
Is it only for non-English character?

**Comment by nnkti18** [comment]
> Is it only for non-English character?

my demo contains only English characters and the keyboard that I used was iOS English default keyboard.

**Comment by harrylefit** [comment]
I also encountered this issue as well. :(

**Comment by singerdmx** [comment]
are you using dev branch or master?

**Comment by singerdmx** [comment]
Oh, it is dev branch. 
Hmm ... weird, I don't have this issue on https://pub.dev/packages/flutter_quill

**Comment by nnkti18** [comment]
> Oh, it is dev branch.
> Hmm ... weird, I don't have this issue on https://pub.dev/packages/flutter_quill

Thanks, I'll have a look at https://pub.dev/packages/flutter_quill.
Btw, I have just created a gif that demonstrate this bug on Zefyr's original example. I edited the content of welcome.note to
`[{"insert":"\n"}]`

![bug_example](https://user-images.githubusercontent.com/27147322/108448289-ffa9a900-7293-11eb-8af2-fef00b79f740.gif)

**Comment by singerdmx** [comment]
Ahh I see
Same issue here
Let me take a look

**Comment by singerdmx** [comment]
Wait, I can no longer repro it using [flutter-quill](https://pub.dev/packages/flutter_quill)
I seem to be able to repro it the very first time, but never again.
Is it a consistent repro for you?

**Comment by singerdmx** [comment]
Even if it occurs, you can use `cut`, right?

**Comment by nnkti18** [comment]
> Wait, I can no longer repro it using [flutter-quill](https://pub.dev/packages/flutter_quill)
> I seem to be able to repro it the very first time, but never again.
> Is it a consistent repro for you?

Yes, you can reproduce it for the first time only. Restart the app to do it again.


> Even if it occurs, you can use `cut`, right?

Cutting is not possible because when I try to long-press the text selection it deselects the current selection (of course I can select the text after the text deselection but cutting and copying are not working on 1.0-dev at the moment). However, when it  
 occurs, I can enter any character to replace the text (deletion is the only key not working).

**Comment by singerdmx** [comment]
I tried restart but cannot repro.
How did you restart?

**Comment by nnkti18** [comment]
> I tried restart but cannot repro.
> How did you restart?

Well, I just use Flutter Hot Restart.

**Comment by singerdmx** [comment]
I tried both using ide and on the phone killing the app

**Comment by nnkti18** [comment]
> I tried both using ide and on the phone killing the app

I tried flutter-quill and I can reproduce it consistently using Flutter Hot restart.

**Comment by singerdmx** [comment]
I was using a real phone, let me try using emulator

**Comment by nnkti18** [comment]
> I was using a real phone, let me try using emulator

You can make a startup screen that navigates to a new screen that has ZefyrEditor, doing so will not require you to restart every time. Just pop and navigate to that screen again

**Comment by singerdmx** [comment]
man I cannot repro on https://pub.dev/packages/flutter_quill
I will ask my teammate to try

**Comment by nnkti18** [comment]
> man I cannot repro on https://pub.dev/packages/flutter_quill
> I will ask my teammate to try


I run flutter_quill on the master channel.
I edited the sample_data.json like this.
`[
   {
      "insert": "\n"
   }
]`

Here is the result
![bug_quill](https://user-images.githubusercontent.com/27147322/108457203-b44bc680-72a4-11eb-921c-922d8f8f0471.gif)

**Comment by li3317** [comment]
Hi I tried looking into this issue as well (using https://pub.dev/packages/flutter_quill), looks like I couldn't repro the bug using a real phone.

**Comment by singerdmx** [comment]
@nnkti18 have you tried using a real phone? Maybe it is an issue with emulator

**Comment by nnkti18** [comment]
> @nnkti18 have you tried using a real phone? Maybe it is an issue with emulator

I reproduced it on a K20 Pro (running Android 10).
A new thing I noticed is that when I try to remove the last character, this log appears
`E/SpannableStringBuilder(30285): SPAN_EXCLUSIVE_EXCLUSIVE spans cannot have a zero length
E/SpannableStringBuilder(30285): SPAN_EXCLUSIVE_EXCLUSIVE spans cannot have a zero length
E/SpannableStringBuilder(30285): SPAN_EXCLUSIVE_EXCLUSIVE spans cannot have a zero length
E/SpannableStringBuilder(30285): SPAN_EXCLUSIVE_EXCLUSIVE spans cannot have a zero length
E/SpannableStringBuilder(30285): SPAN_EXCLUSIVE_EXCLUSIVE spans cannot have a zero length
E/SpannableStringBuilder(30285): SPAN_EXCLUSIVE_EXCLUSIVE spans cannot have a zero length
`

![app](https://user-images.githubusercontent.com/27147322/108466591-722b8080-72b6-11eb-8562-c73b3ebacb47.gif)

**Comment by singerdmx** [comment]
![image](https://user-images.githubusercontent.com/122956/108466864-2c7ca280-7239-11eb-9cb7-7613cff15d4f.png)

I have no hit searching it

https://stackoverflow.com/questions/13670374/android-span-exclusive-exclusive-spans-cannot-have-a-zero-length

It seems it is Android OS issue

**Comment by singerdmx** [comment]
![image](https://user-images.githubusercontent.com/122956/108467022-69e13000-7239-11eb-9934-cc6b13eede11.png)

**Comment by nnkti18** [comment]
It also reproducible on a real iOS device
![bug_ios](https://user-images.githubusercontent.com/27147322/108468687-b1a79c00-72b9-11eb-997a-3c58d6a76dfd.gif)

**Comment by cuongloveit** [comment]
I also have this issue. Could you fix? :((

**Comment by nnkti18** [comment]
@singerdmx hey, can you check the void updateEditingValue(TextEditingValue value) function in editor_input_client_mixin.dart. I think the problem comes from this piece of code
`if (_sentRemoteValues.contains(value)) {
      /// There is a race condition in Flutter text input plugin where sending
      /// updates to native side too often results in broken behavior.
      /// TextInputConnection.setEditingValue is an async call to native side.
      /// For each such call native side _always_ sends an update which triggers
      /// this method (updateEditingValue) with the same value we've sent it.
      /// If multiple calls to setEditingValue happen too fast and we only
      /// track the last sent value then there is no way for us to filter out
      /// automatic callbacks from native side.
      /// Therefore we have to keep track of all values we send to the native
      /// side and when we see this same value appear here we skip it.
      /// This is fragile but it's probably the only available option.
      _sentRemoteValues.remove(value);
      return;
    }`

**Comment by singerdmx** [comment]
I cannot repro your issue. But if you informed what the fix, I can try to apply it and let you try it out.

**Comment by nnkti18** [comment]
> I cannot repro your issue. But if you inform me what the fix is, I can try to apply it and let you try it out.

Okay, I'll let you know when I come up with something new.

**Comment by nnkti18** [comment]
@singerdmx 
![image](https://user-images.githubusercontent.com/27147322/109271190-bd5b0b80-7841-11eb-8b36-65ce7c53005a.png)
I remove this line of code in editor_input_client_mixin.dart
`_sentRemoteValues.add(_lastKnownRemoteTextEditingValue);`
And the issue seems to be fixed but I am not sure what's the consequences

**Comment by singerdmx** [comment]
I can do some testing at our side.
Also you can test it out on this branch https://github.com/singerdmx/flutter-quill/tree/last_char

**Comment by nnkti18** [comment]
> I can do some testing at our side.
> Also you can test it out on this branch https://github.com/singerdmx/flutter-quill/tree/last_char

Thanks, I'll test it.

**Comment by singerdmx** [comment]
![image](https://user-images.githubusercontent.com/122956/109273035-83005880-77c6-11eb-8289-cb922b1cbc8a.png)

`!hasConnection` only happens when you open keyboard for the first time. Later if you hide the keyboard, it is still considered open (attached). 

You want to test `double click a word` that opens keyboard or `using one click` that opens keyboard.

**Comment by singerdmx** [comment]
Also we recently fixed the cursor issue when keyboard is open https://github.com/singerdmx/flutter-quill/commit/a5e1194bb34c8436454e70e6b474dbc1c9fe23df
You might want to check that as well.
Now keyboard won't hide your cursor.

**Comment by singerdmx** [comment]
@nnkti18 I tested and no issue. 
Submitted as https://github.com/singerdmx/flutter-quill/commit/d4b07c5395a5679727b704460ad45766eb959219
You can check it out

**Comment by nnkti18** [comment]
> @nnkti18 I tested and no issue.
> Submitted as [singerdmx/flutter-quill@d4b07c5](https://github.com/singerdmx/flutter-quill/commit/d4b07c5395a5679727b704460ad45766eb959219)
> You can check it out

Thanks

**Comment by pulyaevskiy** [comment]
This should be fixed by #563 and to be released as 1.0-rc.2 later this week.
